lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -701,6 +702,145 @@ void TypeLoweringVisitor::visitExpr(SubindexOp op) {
   opsToRemove.push_back(op);
 }
 
+static IntegerAttr getIntAttr(const APInt &value, MLIRContext *context) {
+  return IntegerAttr::get(IntegerType::get(context, value.getBitWidth()),
+                          value);
+}
+
+// This generates a mux tree for selecting the corresponding element from the
+// flattened vector based on the index.
+// x = a[index] is lowered to ::
+// x = index == 3 ? a_3 : index == 2 ? a_2 : index == 1 ? a_1:a_0
+// TODO: This currently does not handle assignment to vectors.
+void TypeLoweringVisitor::visitExpr(SubaccessOp op) {
+  Value inputVector = op.input();
+  if (inputVector.getType().cast<FIRRTLType>().isa<FlipType>()) {
+    builder->emitError(
+        "Lowering of dynamic assignment to vectors not supported");
+    return;
+  }
+  FIRRTLType vectorElementType = inputVector.getType()
+                                     .cast<FIRRTLType>()
+                                     .cast<FVectorType>()
+                                     .getElementType();
+  // If this is a multi-dim vector, then substitution will happen at the leaf
+  // node, when the non-vector subelement is accessed.
+  if (vectorElementType.isa<FVectorType>()) {
```

Comment:
My concern with this, and the rest of the pass at this point, is that at each op, a complete flattening of the types happen.  This makes each op lowering need to deal with essentially all cases, unless carefully crafted to inject the new ops forward in the module so that the pass will pick them up.  It seems like it would simplify the code if instead of completely flattening the type, you only flattened the outer type, generating code for that.  Then the pass would naturally hit the next layer of flattening when it processed the ops you created.  This would make some operations exist and be immediately deleted, which is bad, but it would ensure that you don't have to deal with vector of vector of bundles with vectors and other arbitrary nesting.  The complete flattening makes sense for instances and arguments, but for normal accessors seems to just cause complexity.